lib/coordinates.rb

```diff
@@ -0,0 +1,12 @@
+class Coordinates
+  attr_reader :value, :grid
+
+  def initialize(value, grid)
+    @value = value
+    @grid = grid
+  end
+
+  def valid?
```

Comment:
Passing in the grid here where it is needed means that the object identity is not tied to the grid object, but you can use the grid here where you need it.

This pattern of receiving an object in an argument and then querying or commanding that object is called the visitor pattern.

Reply:
Interesting!